### ISSUE TITLE
Skale Network Indexer issue

### ISSUE THREAD

**Issue by yohanelly95** [graphprotocol/graph-node#3357]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
![image](https://user-images.githubusercontent.com/33081891/158117092-892aef48-4262-422e-a8e7-e05319fd1c19.png)
When a contract emits only a single event, the indexer seems to index all events emitted properly from the `start block`. No Delegated events are missed (refer screenshot above.)

But in reality, the contract emits many more events that need to be indexed. So now when I add these events, the indexer only indexes blocks from the `current block` and skips all events before that. (does not index from the `start block`). 
TLDR: Indexer skips all events emitted from the start block to the current block.

![code](https://user-images.githubusercontent.com/33081891/158117901-b3d43f43-ab7d-42e5-b113-fd9507da8523.png)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
Deploy subgraph to skale network. Set network as `mainnet` or `skale`. We are using a private RPC provided to us. (There is no issue with the RPC). Add a single event and deploy subgraph. Then add multiple events to the same contract and deploy subgraph.

**What is the expected behavior?**

Contracts that emit multiple events should index all the events from the start block.

**Comment by leoyvens** [comment]
As replied on Discord, this sounds like an issue with the skale eth_getLogs implementation, please try constructing the json-rpc eth_getLogs request for those events combined and sending it manually to the skale node.

**Comment by yohanelly95** [comment]
Tried this as well, I think you're right. When there is more than 1 topic, the `eth_getLogs` fails and does not return any logs. 

With 1 topic: `curl https://dappnet-api.skalenodes.com/v1/ \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address": "0x1Be704ddCdb50703a953139dd505729903891A22", "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"] }],"id":0}'`

The above returns logs no problems. 

But if the request has more than 1 topic, logs are not returned (get an empty response): 

`curl https://dappnet-api.skalenodes.com/v1/ \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address": "0x1Be704ddCdb50703a953139dd505729903891A22", "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5", "0x736990cb4424cb3b541e24055819082dfdec9a3c74c5b7b771350f0cff7ec195"] }],"id":0}'`

We also tried requesting for logs with no topics, this works as well and returns all logs. Reference: 

`curl https://dappnet-api.skalenodes.com/v1/faint-acubens \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address": "0x1Be704ddCdb50703a953139dd505729903891A22"}],"id":0}'`

This request without `topics` returns all logs for that contract address. 

How can I resolve this?

**Comment by dimalit** [comment]
@yohanelly95 Could you please provide an example of event that should be returned by the getLogs request that you posted before? I just made few tests - and events with 2 topics seem to be searched successfully. So please provide exact event-request pair, where the "request" should successfully return the "event" but it does not do it.

**Comment by yohanelly95** [comment]
So this is the response with a single topic : `{
    "id": 0,
    "jsonrpc": "2.0",
    "result": [{
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xc61ecb4ac387b137e7678376cf2ef56c920d5418cb7825b57a806af3770dc1df",
        "blockNumber": "0x1c4818",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ec000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000062221e8c000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000426aa00000000000000000000000000000000000000000000000000000000003f17cc",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x73a3b02d6b0256af424044f8e3c5c74edcb60198348e42e929b95c50dc304ea0",
        "transactionIndex": "0x0",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xc61ecb4ac387b137e7678376cf2ef56c920d5418cb7825b57a806af3770dc1df",
        "blockNumber": "0x1c4818",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ec000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000062221e8c000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000426aa00000000000000000000000000000000000000000000000000000000003f17cc",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0xa5e9114d470a4bbaad9e09774ee0356f2b65c7671e6c4b30119350e32c99cf2f",
        "transactionIndex": "0x1",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xdf5454411aa1e5b6fbd1c1604972c46d6f0ca298135e8265d9a9e23cb2f29c16",
        "blockNumber": "0x1c48c2",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ed000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000006222259700000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000041ffc00000000000000000000000000000000000000000000000000000000003f1c3e",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x7bb3e5a64926d407fa6080cee58d5b4ed433129742584c8de0b4c7e45b958d4c",
        "transactionIndex": "0x0",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xdf5454411aa1e5b6fbd1c1604972c46d6f0ca298135e8265d9a9e23cb2f29c16",
        "blockNumber": "0x1c48c2",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ed000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000006222259700000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000041ffc00000000000000000000000000000000000000000000000000000000003f1c3e",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x6c5940258c1c5f97d40d052e9e44d9e8ba902d5b51bb8482b1ae8a727ab58636",
        "transactionIndex": "0x1",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0x3c4a72d579aeb43928994056a52fb5d8e1942a97380c250786620192c485172a",
        "blockNumber": "0x1c496c",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ee000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000062222c9f0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000004214b00000000000000000000000000000000000000000000000000000000003eb19b",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0xbf2f97a71af7de23ec19b228c7d769eca7e35ffdaad0a1626577ec70927e1ad3",
        "transactionIndex": "0x0",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0x36efba87508231aa455f075b34f9429ae0dbc321e7b0f30bb2c23074ebae2479",
        "blockNumber": "0x1c496d",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ee000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000062222ca00000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000004214b00000000000000000000000000000000000000000000000000000000003eb19b",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x52bd300cc04cfc63ef6e0f47a204a89f2369859cedcf3f86f8ca60e4e519b305",
        "transactionIndex": "0x0",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xb605cb6b33de80c759158ac93ba3722f561be496a26e0625b44850548d38f0c6",
        "blockNumber": "0x1c4a15",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ef000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000622233a60000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000004154900000000000000000000000000000000000000000000000000000000003e2ea0",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x92effb179f51af648e01c4973faccc7765dc9d82a798be240a97e5f21d830e5a",
        "transactionIndex": "0x0",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0xb605cb6b33de80c759158ac93ba3722f561be496a26e0625b44850548d38f0c6",
        "blockNumber": "0x1c4a15",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4ef000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000622233a60000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000004154900000000000000000000000000000000000000000000000000000000003e2ea0",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x6361ba4145f0f22745c8e16ae3a96578ae0346b919452f22a1b0eff579438956",
        "transactionIndex": "0x1",
        "type": "mined"
    }, {
        "address": "0x1be704ddcdb50703a953139dd505729903891a22",
        "blockHash": "0x843d66494283aa2179abe3d72e8234fcdd461d890bb12ca1713f2a1120b42c61",
        "blockNumber": "0x1c4abe",
        "data": "0x00000000000000000000000000000000000000000000000000000000000df4f0000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000062223aa80000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000004131800000000000000000000000000000000000000000000000000000000003e2275",
        "logIndex": "0x0",
        "polarity": true,
        "topics": ["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5"],
        "transactionHash": "0x5ee1ea7e7cb5bdbc2ac214fdc6be2291e51dd1489da0a79a36fc353a259fa475",
        "transactionIndex": "0x0",
        "type": "mined"
    },
]}`

This is the response I get when I query eth_getLogs with more than 1 topic:

`{
    "id": 0,
    "jsonrpc": "2.0",
    "result": []
}`

Note: add `/faint-acubens` to the RPC url.

**Comment by dimalit** [comment]
@yohanelly95 From your example events it follows that the response you get when requesting TWO topics is correct. When specifying TWO topics you are trying to request events that contain BOTH of them. And clearly there are no such. See "A note on specifying topic filters:" https://eth.wiki/json-rpc/API#a-note-on-specifying-topic-filters

**Comment by leoyvens** [comment]
@dimalit right, the request @yohanelly95 posted would be an AND. However if you nest the array, to make an OR on the first topic, the response is still empty.
```
{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address": "0x1Be704ddCdb50703a953139dd505729903891A22", "topics": [["0x60b28e9913876d5cd18aedee156a4b9a83cb2e4661a47a60d5dbe1f6f71f4bf5", "0x736990cb4424cb3b541e24055819082dfdec9a3c74c5b7b771350f0cff7ec195"]] }],"id":0}
```

**Comment by yohanelly95** [comment]
You are right @dimalit, I have now tried two topics emitted from two contracts that should exist together. I tried A & B as well as A or B and both do not return any logs. Please refer to this link -https://reqbin.com/c-celwiuy9.

If there is something that doesn't quite add up, please refer to this txn hash - http://faint-acubens.explorer.dappnet.skalenodes.com/tx/0x9df3dad4f227607600fb088aef40c80fee35830f23cbf1b1a87ddd5df29ad510/logs  cc @leoyvens

**Comment by dimalit** [comment]
@yohanelly95 
1) Please use "0x" prefix when specifying topics.
2) I just checked out (based on your example from reqbin), and found out that the request with (single) topic "0x272..." successfully returns some results. But request for "0xd25.." topic gives 0 results. Please ensure that there really are logs with this topic.
3) Indeed, we confirmed that "OR" requests are broken on skale-chains. We included this problem into the plan for 2.1 release, o stay tuned. Meanwhile you (possibly) can use 2 separate requests. https://github.com/skalenetwork/skaled/issues/1030

**Comment by yohanelly95** [comment]
I see, at the moment we cannot use 2 separate requests, since the graph indexer does this internally. If we index any contract with more than 1 event, all events from the start block to the current block are lost. Data integrity as lost when we redeploy indexer. This is a blocker for us at the moment.

**Comment by yohanelly95** [comment]
We have figured out a solution for now, we have removed the contract address from the subgraph.yaml. So now eth_getLogs will return all events, we then add BlockManager/Staker Manager address checks in the mapping directly. So for any event to get indexed, the event signature AND the contract address needs to match. This is a workaround until eth_getLogs is fixed from Skales end. 
![Screenshot 2022-03-30 at 2 14 30 PM](https://user-images.githubusercontent.com/33081891/160790336-e062338d-0233-4fb5-b6d6-1147e48bbc25.png)
